openss7-0.9.2.D.SLES9_K2.6.5_rc1(1)/openss7-0.9.2.D.SLES9_K2.6.5_rc1/LiS/man/man4/fifo.4s.man
'\" rtp
.\" -*- nroff -*- vim: ft=nroff nocin nosi
.\"
.\" @(#) fifo.4s.man,v 1.1.4.3 2006/04/05 09:42:35 brian Exp
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2001-2006  OpenSS7 Corporation <http://www.openss7.com/>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.  The author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified 2006/04/05 09:42:35 by brian
.\"
.\" -------------------------------------------------------------------------
.\"
.\" fifo.4s.man,v
.\" Revision 1.1.4.3  2006/04/05 09:42:35  brian
.\" - moved .so line and updated headers
.\"
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database lis.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
.R2
.so lis.macros
.\"
.\"
.TH FIFO 4S "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Devices"
.\"
.\"
.SH NAME
.PP
.B fifo
\- pseudo-driver for \fISTREAMS\fR \fBfifos\fR
.\"
.\"
.SH SYNOPSIS
.PP
.\"
.\"
.SH DESCRIPTION
.PP
The
.B fifo
pseudo-driver
(which is internal to
.B LiS\c
)
.[
LiS
.]
provides
.I STREAMS
.I fifos
as single character special files, and
.I STREAMS
.I pipes
as pairs of character special files which are interconnected (see
.BR pipe "(3)).  "
.sp
.I STREAMS
.I fifos
differ from typical
.I STREAMS
character special files in that there are not separate
.I "stream head"
and
.I driver
queue pair within the
.I STREAMS
file.  Instead, a
.I fifo
is created with only a single queue pair for the
.IR "stream head" .
Moreover, in a typical
.I driver
queue pair, the write queue is not connected to a next queue.
In a
.IR fifo ,
the write queue is directed to the read queue of the pair.
A
.I pipe
comprises a pair of
.IR fifos ", "
with the write queue of each pair directed to the read queue of the other.
The two
.B fifos
comprising a
.B pipe
are referred to as
.IR peers ", "
and each somewhat represents a
.I driver
to the other.
As a degenerate case, a
.B fifo
is its own peer.
.sp
.I STREAMS
.I modules
may be pushed onto
.B fifos
and
.BR pipes ", "
but should not expect a
.I driver
below them; instead, the
.BR SAMESTR "()"
function should be used from the write queue of a pair to determine if
the module is the lowest in the
.I STREAMS
file (this is called the
.IR midpoint "). "
The structure of a
.B fifo
or
.B pipe
is preserved when modules are pushed (and popped); i.e., the write
queue at the
.I midpoint
will always be directed at the read queue of the
.IR peer "."
.sp
Input and output are handled at a
.B fifo
stream head as they would normally be handled
at a stream head.  In
LiS,
.[
LiS
.]
an
.BR "fifo open" "() "
entry point exists to assign minor device numbers to new opens under the
.B fifo
major device number, and a
.BR close "() "
entry point is used correspondingly to release them.  These functions
are kept in a
.I streamtab
data structure (as they would normally be for any
.I STREAMS
driver or module) which is private to the
LiS
.[
LiS
.]
implementation.
.\"
.\"
.SH "APPLICATION USAGE"
.PP
In the current Linux kernels, character special major numbers are limited
to 16 bits, and major and minor device numbers to 8 bis each.  This limits
a system to 256 total major device numbers and 256 total minor devices
per major device number.  This is a rather severe limitation where
mechanisms like
.B fifos
and
.B pipes
are concerned.
.sp
However, a driver may handle more than one major device number.  The
.B fifo
pseudo-driver uses this to overcome this limitation, by supporting the
automatic allocation and use of multiple major device numbers for
.B fifos
and
.BR pipes .
Specifying more than 256 minor devices is done in the usual manner, i.e.,
by specifying the number of "units" in the appropriate
.B Config
file.  Enough major device numbers will be allocated to cover the
requested number of minor devices (if available, else an error will
occur in 
.BR strconf "(8)). "
The number allocated will include one minor device per major number
to be used as a
.BR fifo "\-specific"
.I clone
minor device (specifically, minor number 0), which exhibits special behavior.
.sp
Normally, when cloning is done via the
.B clone
pseudo-driver, the
.B clone
major device number is used, along with the desired actual major number
as the minor device number.  When an
.BR open "() "
is performed on such a device, the
.BR "clone open" "() "
routine in turn calls the appropriate driver's
.BR open "(), "
with the
.I sflag
parameter set to
.BR CLONEOPEN .
The driver's
.BR open "() "
is expected in this case to allocate an unused minor device number,
and return it via an entirely new device number in the
.I "devp" 
parameter.  In this way, a driver can change the device number to be
used for a
.I STREAMS
file.
When minor device 0 for a specified for a
.B fifo
major device, the driver will also clone a new minor device number.
However,
LiS
.[
LiS
.]
opens
.B fifo
devices differently; specifically, when an already-opened
.BR fifo "\-specific"
clone minor device is reopened, the new and subsequent opens will use
the already-opened clone.
Thus, using minor device 0 for a
.B fifo
when creating a file system node will ensure that all concurrent
opens of the associated path name will use the same
.I STREAMS
file; at the same time, opens of different file system nodes via
different paths will open their respectively different
.I STREAMS
files.
This is essentially how kernel
.B fifos
behave -applications and users of
.I STREAMS
.B fifos
don't have to keep track of minor numbers to achieve this same behavior
when it is desired.
.sp
It is in fact recommended that only two forms of file system nodes be used for
.I STREAMS
.BR fifos ": "
the
.B clone
major number as major number with a
.B fifo
major number as minor number, to be used when
.I every
open of the associated path must clone a new
.BR fifo ", "
and a
.B fifo
major number as major number with 0 as the minor number, to be used
when new opens are to clone a new
.B fifo
but subsequent concurrent opens are to use the already opened
.BR fifo ". "
These are represented by two device special file paths created when LiS
.[
LiS
.]
is installed:
.RB "/" dev "/" fifo
for the former, and
.RB "/" dev "/" fifo.0
for the latter.  It is recommended that these be used, possibly
along with the equivalent of
.BR stat "(2) "
to determine appropriate major device numbers for the
.B clone
and
.B fifo
pseudo-drivers, which are also determined when LiS
.[
LiS
.]
is installed.
It can be noted that
.B pipes
are actually created as instances of the former, after which the write
queues are peer-connected.
.sp
The
.B fifo
pseudo-driver allocates minor devices in
.I round-robin
fashion; i.e., a list of available minor devices is kept, and once a
minor number is finally closed, it is put at the end of this list.
Thus, a
.B fifo
minor device which is opened and closed will not be immediately reused.
.\"
.\"
.SH WARNINGS
.PP
Because
.I STREAMS
.B fifos
and
.B pipes
are implemented as character special devices, they do not appear as
.B pipe
devices when examined with
.BR stat "(2) "
or the equivalent (e.g., 
.BR ls "(1)); "
i.e. the
.B S_IFIFO
indication is not set in the mode -
.B S_IFCHR
is set instead, and the actual device number is indicated in the
.I st_rdev
field of the
.I stat
data structure.
.sp
Because of the potential use of multiple major numbers, applications
should not depend on a
.B fifo
or
.B pipe
having a specific major device number, nor should an application depend on
all
.B fifos
and
.B pipes
having the same major device number.
.\"
.\"
.SH DEVICES
.PP
.\"
.\"
.SH "SEE ALSO"
.PP
.BR clone "(4), "
.BR connld "(4), "
.BR fifo "(4), "
.BR ls "(1), "
.BR pipe "(3), "
.BR pipemod "(4), "
.BR STREAMS "(4), "
.BR stat "(2), "
.BR strconf "(8)"
.\"
.\"
.SH HISTORY
.PP
UNIX\(rg\ System V Release 4 (SVR4)
.[
svr4
.]
.\"
.\"
.SH AUTHOR
.PP
John Boyd, protologos LLC.
.[
$LIST$
.]
.T3


openss7-0.9.2.D.SLES9_K2.6.5_rc1(1)/openss7-0.9.2.D.SLES9_K2.6.5_rc1/streams/src/drivers/fifo.c
/*****************************************************************************

 @(#) fifo.c,v OpenSS7-0_9_2_D_rc2(0.9.2.39) 2006/06/14 10:37:21

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2006  OpenSS7 Corporation <http://www.openss7.com>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software; you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation; version 2 of the License.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License along with
 this program; if not, write to the Free Software Foundation, Inc., 675 Mass
 Ave, Cambridge, MA 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified 2006/06/14 10:37:21 by brian

 *****************************************************************************/

#ident "@(#) fifo.c,v OpenSS7-0_9_2_D_rc2(0.9.2.39) 2006/06/14 10:37:21"

static char const ident[] =
    "fifo.c,v OpenSS7-0_9_2_D_rc2(0.9.2.39) 2006/06/14 10:37:21";

#include <linux/config.h>
#include <linux/version.h>
#include <linux/module.h>
#include <linux/init.h>

#include <asm/semaphore.h>

#include <sys/stream.h>
#include <sys/strconf.h>
#include <sys/strsubr.h>
#include <sys/ddi.h>

#ifndef HAVE_KFUNC_MODULE_PUT
#define module_refcount(__m) atomic_read(&(__m)->uc.usecount)
#endif

#include "sys/config.h"
#include "src/kernel/strsched.h"	/* for allocstr */
#include "src/kernel/strsad.h"	/* for autopush */
#include "src/modules/sth.h"
#include "fifo.h"		/* extern verification */

#define FIFO_DESCRIP	"UNIX SYSTEM V RELEASE 4.2 FAST STREAMS FOR LINUX"
#define FIFO_COPYRIGHT	"Copyright (c) 1997-2006 OpenSS7 Corporation.  All Rights Reserved."
#define FIFO_REVISION	"LfS fifo.c,v OpenSS7-0_9_2_D_rc2(0.9.2.39) 2006/06/14 10:37:21"
#define FIFO_DEVICE	"SVR 4.2 STREAMS-based FIFOs"
#define FIFO_CONTACT	"Brian Bidulock <bidulock@openss7.org>"
#define FIFO_LICENSE	"GPL"
#define FIFO_BANNER	FIFO_DESCRIP	"\n" \
			FIFO_COPYRIGHT	"\n" \
			FIFO_REVISION	"\n" \
			FIFO_DEVICE	"\n" \
			FIFO_CONTACT	"\n"
#define FIFO_SPLASH	FIFO_DEVICE	" - " \
			FIFO_REVISION	"\n"

#ifdef CONFIG_STREAMS_FIFO_MODULE
MODULE_AUTHOR(FIFO_CONTACT);
MODULE_DESCRIPTION(FIFO_DESCRIP);
MODULE_SUPPORTED_DEVICE(FIFO_DEVICE);
MODULE_LICENSE(FIFO_LICENSE);
#if defined MODULE_ALIAS
MODULE_ALIAS("streams-fifo");
#endif
#endif

#ifndef CONFIG_STREAMS_FIFO_NAME
//#define CONFIG_STREAMS_FIFO_NAME "fifo"
#error "CONFIG_STREAMS_FIFO_NAME must be defined."
#endif
#ifndef CONFIG_STREAMS_FIFO_MODID
//#define CONFIG_STREAMS_FIFO_MODID 6
#error "CONFIG_STREAMS_FIFO_MODID must be defined."
#endif
#ifndef CONFIG_STREAMS_FIFO_MAJOR
//#define CONFIG_STREAMS_FIFO_MAJOR 0
#error "CONFIG_STREAMS_FIFO_MAJOR must be defined."
#endif

modID_t modid = CONFIG_STREAMS_FIFO_MODID;

#ifndef module_param
MODULE_PARM(modid, "h");
#else
module_param(modid, ushort, 0);
#endif
MODULE_PARM_DESC(modid, "Module identification number for STREAMS-based FIFOs");

#ifdef MODULE_ALIAS
MODULE_ALIAS("streams-modid-" __stringify(CONFIG_STREAMS_FIFO_MODID));
MODULE_ALIAS("streams-driver-fifo");
#endif

major_t major = CONFIG_STREAMS_FIFO_MAJOR;

#ifndef module_param
MODULE_PARM(major, "h");
#else
module_param(major, uint, 0);
#endif
MODULE_PARM_DESC(major, "Major device number for STREAMS-based FIFOs.");

#ifdef MODULE_ALIAS
MODULE_ALIAS("char-major-" __stringify(CONFIG_STREAMS_FIFO_MAJOR) "-*");
MODULE_ALIAS("/dev/fifo");
#ifdef LFS
MODULE_ALIAS("streams-major-" __stringify(CONFIG_STREAMS_FIFO_MAJOR));
MODULE_ALIAS("/dev/streams/fifo");
MODULE_ALIAS("/dev/streams/fifo/*");
#endif
#endif

static struct module_info fifo_minfo = {
	.mi_idnum = CONFIG_STREAMS_FIFO_MODID,
	.mi_idname = CONFIG_STREAMS_FIFO_NAME,
	.mi_minpsz = STRMINPSZ,
	.mi_maxpsz = STRMAXPSZ,
	.mi_hiwat = STRHIGH,
	.mi_lowat = STRLOW,
};

static struct qinit fifo_rinit = {
	.qi_putp = strrput,
	.qi_qopen = str_open,
	.qi_qclose = str_close,
	.qi_minfo = &fifo_minfo,
};

static struct qinit fifo_winit = {
	.qi_putp = strwput,
	.qi_srvp = strwsrv,
	.qi_minfo = &fifo_minfo,
};

static struct streamtab fifo_info = {
	.st_rdinit = &fifo_rinit,
	.st_wrinit = &fifo_winit,
};

/* 
 *  -------------------------------------------------------------------------
 *
 *  INITIALIZATION
 *
 *  -------------------------------------------------------------------------
 */

static struct cdevsw fifo_cdev = {
	.d_name = CONFIG_STREAMS_FIFO_NAME,
	.d_str = &fifo_info,
	.d_flag = D_MP,
	.d_fop = &strm_f_ops,
	.d_mode = S_IFIFO | S_IRUGO | S_IWUGO,
	.d_kmod = THIS_MODULE,
};

/* 
 *  -------------------------------------------------------------------------
 *
 *  Special open for fifo devices.
 *
 *  -------------------------------------------------------------------------
 */

/*
 *  fifo_open: - open a fifo device node
 *  @inode: the external filesystem inode
 *  @file: the external filesystem file pointer
 *
 *  fifo_open() is only used to open a fifo device (named pipe) from a character device node in an
 *  external filesystem.  This is never called for direct opens of a specfs device node (for direct
 *  opens, see spec_dev_open() in strspecfs.c).  The character device inode is opened directly and
 *  no inode in the shadow filesystem is addressed.
 */
STATIC int
fifo_open(struct inode *inode, struct file *file)
{
	int err;
	dev_t dev = makedevice(fifo_cdev.d_modid, 0);

	_printd(("%s: %s: putting file operations\n", __FUNCTION__, file->f_dentry->d_name.name));
	_printd(("%s: %s: getting file operations\n", __FUNCTION__, fifo_cdev.d_name));
	{
		struct file_operations *f_op;

		err = -ENXIO;
		if (!(f_op = fops_get(fifo_cdev.d_fop))) {
			_ptrace(("Error path taken!\n"));
			goto error;
		}
#ifdef CONFIG_STREAMS_DEBUG
		if (f_op->owner)
			_printd(("%s: [%s] new f_ops count is now %d\n", __FUNCTION__,
				 f_op->owner->name, module_refcount(f_op->owner)));
		else
			_printd(("%s: new f_ops have no owner!\n", __FUNCTION__));
#endif
#ifdef CONFIG_STREAMS_DEBUG
		if (file->f_op->owner)
			_printd(("%s: [%s] old f_ops count is now %d\n", __FUNCTION__,
				 file->f_op->owner->name, module_refcount(file->f_op->owner) - 1));
		else
			_printd(("%s: old f_ops have no owner!\n", __FUNCTION__));
#endif
		fops_put(file->f_op);
		file->f_op = f_op;
	}
	file->private_data = &dev;	/* use this device number instead of inode number */
	file->f_flags &= ~O_CLONE;	/* FIFOs never clone */
	err = file->f_op->open(inode, file);
      error:
	return (err);
}

STATIC struct file_operations fifo_f_ops ____cacheline_aligned = {
	.owner = NULL, /* yes NULL */
	.open = &fifo_open,
};

/* 
 *  -------------------------------------------------------------------------
 *
 *  REGISTRATION
 *
 *  -------------------------------------------------------------------------
 */

#ifdef CONFIG_STREAMS_FIFO_OVERRIDE
static const struct file_operations *_def_fifo_ops =
    ((typeof(_def_fifo_ops)) HAVE_DEF_FIFO_OPS_ADDR);

static struct file_operations fifo_tmp_ops;

STATIC void
register_fifo(void)
{
	fifo_tmp_ops = *_def_fifo_ops;
	_def_fifo_ops->owner = fifo_f_ops.owner;
	_def_fifo_ops->open = fifo_f_ops.open;
}
STATIC void
unregister_fifo(void)
{
	_def_fifo_ops->open = fifo_tmp_ops.open;
	_def_fifo_ops->owner = fifo_tmp_ops.owner;
}
#endif

/* 
 *  -------------------------------------------------------------------------
 *
 *  INITIALIZAION
 *
 *  -------------------------------------------------------------------------
 */

#ifdef CONFIG_STREAMS_FIFO_MODULE
static
#endif
int __init
fifo_init(void)
{
	int err;

#ifdef CONFIG_STREAMS_FIFO_MODULE
	printk(KERN_INFO FIFO_BANNER);
#else
	printk(KERN_INFO FIFO_SPLASH);
#endif
	fifo_minfo.mi_idnum = modid;
	if ((err = register_cmajor(&fifo_cdev, major, &fifo_f_ops)) < 0)
		return (err);
#ifdef CONFIG_STREAMS_FIFO_OVERRIDE
	register_fifo();	/* This is safe */
#endif
	if (major == 0 && err > 0)
		major = err;
	return (0);
};

#ifdef CONFIG_STREAMS_FIFO_MODULE
static
#endif
void __exit
fifo_exit(void)
{
#ifdef CONFIG_STREAMS_FIFO_OVERRIDE
	unregister_fifo();	/* This is not safe... */
#endif
	unregister_cmajor(&fifo_cdev, major);
};

#ifdef CONFIG_STREAMS_FIFO_MODULE
module_init(fifo_init);
module_exit(fifo_exit);
#endif
